### DISCUSSION TITLE
nuxt usefetch transform with typescript

### DISCUSSION THREAD

**Discussion by Fanreza** [nuxt/nuxt#21248] [Questions] [answered]
```js
const {
	data: dataProduct,
	error: errorProduct,
	pending: loadingProduct,
} = await useFetch<ProductModel>("/products/list/", {
	transform(data) {
		return data.data.map((item): string => {
			return item.coverUrl || "/broken.png";
		});
	},

	server: false,
});
```

why i got error ``Type '(data: ProductModel) => string[]' is not assignable to type '_Transform<ProductModel, ProductModel>'.
  Type 'string[]' is missing the following properties from type 'ProductModel': status, data``

**Comment by warflash** [comment]
What does your `ProductModel` type look like?

**Comment by Fanreza** [comment]
```js
export interface ProductModel {
	status: string;
	data: ProductDataModel[];
}

export interface ProductDataModel {
	productId: string;
	title: string;
	coverPicture?: string;
	coverType?: string;
	coverUrl?: string;
	coverThumbnail?: string;
	description?: string;
	priceCurrency: string;
	price: number;
	language: string;
	discount?: number;
	productTypeId: number;
	folder?: string;
	externalId?: string;
	productCode?: string;
	productStatus: string;
	pinned: boolean;
	comingSoon: boolean;
	ProductType: ProductType;
	totalModules: number;
	ratingProduct: number;
}

export interface ProductType {
	productTypeId: number;
	name: string;
	cover?: string;
	description?: string;
	language?: string;
}
```

it looks like i cant transform it and return as array of string

**Comment by tom-velir** [comment, marked as answer]
I solved this by omitting the generic from the `useFetch` call and just assigning the type of the response to the `transform` param. You might try something similar:

```ts
const { data, error, pending } = await useFetch("/products/list/", {
  transform(data: ProductModel) {
    return data.data.map((item): string => {
      return item.coverUrl || "/broken.png";
    });
  },
  server: false,
});
```

**Reply by Fanreza** [reply]
wow thanks man it works, and also the return type now following the transform return type

**Reply by LeonardoRick** [reply]
@Fanreza is it? My `data` type is being settle to `globalThis.Ref<any>`. Am I missing something?

**Comment by shao-wang-me** [comment]
You can resolve the issue by defining your own wrapper function around `useFetch()`. This allows separate input and output types for the `transform` option without resorting to type assertions (`as`) or omitting type annotations:

```typescript
import type { UseFetchOptions } from '#app'

function useAPI<Raw, Data = Raw>(
  url: string,
  options: UseFetchOptions<Raw, Data> = {}
) {
  return useFetch(url, options)
}
```

**Comment by HardMachinesLab** [comment]
Hi everyone ! 

**I encountered the same issue** _and initially worked around it by removing the TypeScript generics_, [...]  **But I wasn’t fully satisfied with that fix**. After digging deeper, I realized that :


>[!important]
>`transform` in `useFetch` is **not** a “map type” utility— it’s intended to post‑process or enrich the data **without** changing its shape. **According to the docs, its signature is:**
> ```ts
>  transform?: (input: DataT) => DataT | Promise<DataT>
> ```
> 🟥  This means the **input** and **output** of your `transform` callback **must** both be of the same `DataT`. 

### For example, the following compiles without errors:

<img width="725" height="287" alt="fetchTechinque014" src="https://github.com/user-attachments/assets/e9b04925-99f2-422d-8dc4-d45bce5f941b" />

### Back to the task at hand 
```ts
const {
	data: dataProduct,
	error: errorProduct,
	pending: loadingProduct,

        // 1️⃣  ResT = ProductModel
} = await useFetch<ProductModel>("/products/list/", {
        // 2️⃣   infered type is ProductModel ⚠️ 
	transform(data) {
		return data.data.map((item): string => {
			return item.coverUrl || "/broken.png";
		});
	},

	server: false,
});   // → you are try to return string[]
```
>[!caution]
> **This does not align to the transform signature**: `transform?: (input: DataT) => DataT | Promise<DataT>` . Soo
> `Type '(data: ProductModel) => string[]' is not assignable to type '_Transform<ProductModel, ProductModel>'. Type 'string[]' is missing the following properties from type 'ProductModel': status, data`


It also explains why attempting to do something like:

```ts
// ❓ **What's is the `typeof item` ? --> infered as `any`
return data.data.map((item): string => item.coverUrl || '/broken.png')
```

will never compile under `useFetch< ProductModel>`, since you’re returning an array of strings instead of the original ` ProductModel` shape.

>[!tip]
>**An alternative is to mais your transformation under `watchEffet` for `compute`**

Happy Coding !!!

>**Hard Machine × 🤖 Aegis**
